Edit 2020 Markup chapter (#1435)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/markup.md

title: Markup
description: Markup chapter of the 2020 Web Almanac covering general observations, the use of elements and attributes, as well as trivia and trends.
authors: [j9t, catalinred, iandevlin]
j9t_bio: Jens Oliver Meiert is a web developer and author (<a href="https://leanpub.com/css-optimization-basics"><cite>CSS Optimization Basics</cite></a>, <a href="https://leanpub.com/web-development-glossary"><cite>The Web Development Glossary</cite></a>), who works as an engineering manager at <a href="https://www.jimdo.com/">Jimdo</a>. He's an expert on web development where he specializes in HTML and CSS optimization. Jens contributes to technical standards and regularly writes about his work and research, particularly on his website, <a href="https://meiert.com/en/">meiert.com</a>.
catalinred_bio: Catalin Rosu is a front-end developer at <a href="https://www.caphyon.com/">Caphyon</a> and currently works on <a href="https://www.wattspeed.com/">Wattspeed</a>. He has a passion for web standards and a keen eye for great UX & UI, things he <a href="https://twitter.com/catalinred">tweets</a> and writes about on <a href="https://catalin.red/">his website</a>.

> # iandevlin_bio: @@

Author: Added!

![image](https://user-images.githubusercontent.com/1120896/98471273-13377500-21b9-11eb-8f0a-5675c4a8e1d4.png)

reviewers: [zcorpan, matuzo, bkardell]
analysts: [Tiggerito]
translators: []
discuss: 2039
results: https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/
queries: 03_Markup
featured_quote: We approach near-complete adoption of living HTML, are quick to prune our pages of fads, and we're fast in adopting and shunning frameworks. And still, there are no signs we exhausted the options HTML gives us.
featured_stat_1: 85.73%
featured_stat_label_1: Percentage of pages that use the "living" HTML doctype
featured_stat_2: 30,073

> featured_stat_label_2: Number of non-standard `h7` elements

Reviewer: Only HTML and not Markdown in featured quotes at present (will see if I can change in future).

```suggestion
featured_stat_label_2: Number of non-standard <code>h7</code> elements
```

> featured_quote: We approach near-complete adoption of living HTML, are quick to prune our pages of fads, and we're fast in adopting and shunning frameworks. And still, there are no signs we exhausted the options HTML gives us.

Reviewer: “Wrong” is a incorrect. “Not as elegant” definitely has some merit.

We are using the Google Developer Style Guide for the Web Almanac and [it recommends straight quote marks](https://developers.google.com/style/quotation-marks#straight-and-curly-quotation-marks):

> __Straight and curly quotation marks__
>
> Most typefaces support two forms of quotation marks and apostrophes: straight marks and curly, or typographic, marks. Some tools, like Google Docs, automatically convert straight quotation marks and apostrophes to the curly versions as you type. However, our guidance is to always use straight quotation marks and straight apostrophes, for these reasons:
> 
> - It's easy to get the direction of curly quotation marks (especially apostrophes) wrong. Using straight marks avoids this problem.
> - Code requires straight marks.
> - Many writers use coding tools (for example, Subversion) for working directly with markup. Because these tools do not produce curly marks by default, the writer must manually insert the marks, which is error prone.

So this is a conscious decision to follow this practice. I understand you both disagree with this convention and decision, but that is what we have gone with for the 2019 Web Almanac and that is what is listed in our [Editor’s Guide](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Editors'-Guide#suggested-process). We need to be consistent here.

I suggest you [raise a new issue for discussion](https://github.com/HTTPArchive/almanac.httparchive.org/issues) as to the pros and cons for this (and the effort to convert the 2019 Web Almanac!) but for now, and for this pull request, we should follow the convention already set and style guide we are using.

> featured_quote: We approach near-complete adoption of living HTML, are quick to prune our pages of fads, and we're fast in adopting and shunning frameworks. And still, there are no signs we exhausted the options HTML gives us.

Author: Moving this discussion to https://github.com/HTTPArchive/almanac.httparchive.org/issues/1485

<p class="note">
  In this Markup chapter, we're focusing almost exclusively on HTML, rather than SVG or MathML, which are also considered markup languages. Unless otherwise noted, stats presented in this chapter refer to the set of mobile pages. Additionally, the data for all Web Almanac chapters is open and available. Take a look at <a href="https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/">the results</a> and <a href="https://discuss.httparchive.org/t/2039">share your observations</a> with the community!
</p>

## General

In this section, we're covering the higher-level aspects of HTML like document types, the size of documents, as well as the use of comments and scripts. "Living HTML" is very much alive!

### Doctypes

{{ figure_markup(
  caption="Percent of pages with a doctype.",

>   content="96.82%",

Reviewer 2: I think it's more interesting that 85.73% of pages use the `<!DOCTYPE html>` doctype.

Author: I think it's good to have a figure for each stat, so there's something people can permalink in references. 96.82% is only presented in the text, so that's why I gave it a standalone "big number" figure.

Do you also think it's worth highlighting 85.73% as a big number in addition to the table below?

<figcaption>{{ figure_link(caption="The 5 most popular doctypes.", sheets_gid="1981441894", sql_file="summary_pages_by_device_and_doctype.sql") }}</figcaption>
</figure>

You can already tell how the numbers decrease quite a bit after XHTML 1.0, before entering the long tail with a few standard, some esoteric, and also bogus doctypes.

Two things stand out from these results:

1. Almost 10 years after [the announcement of living HTML](https://blog.whatwg.org/html-is-the-new-html5) (aka "HTML5"), living HTML has clearly become the norm.
2. The web before living HTML can still be seen in the next most popular doctypes, like XHTML 1.0. XHTML. Although their documents were likely delivered as HTML with a MIME type of `text/html`, these older doctypes are not dead yet.

### Document size

A page's document size refers to the amount of HTML bytes transferred over the network, including compression if enabled. At the extremes of the set of 6.3 million documents:

* 1,110 documents are empty (0 bytes).
* The average document size is 50.35 KB (in most cases [compressed](https://w3techs.com/technologies/details/ce-gzipcompression)).

> * The largest document by far weighs 64.16 _MB_, almost deserving its own analysis and chapter in the Web Almanac.

Author: @Tiggerito could you confirm the correct stat for "largest document"?

> * The largest document by far weighs 64.16 _MB_, almost deserving its own analysis and chapter in the Web Almanac.

Reviewer 3: I think the real crawl came up with 64MB for that page, but a subsequent test showed that had reduced it. I also think the 64MB is for all html resources on the page, not just the initial file.

https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/edit?ts=5f4e700c#gid=813852329

I just noticed in the data that there are a few pages with exactly the same large bytesHtmlDoc of 16,777,215.

I tested one at random. It took over a minute to load 22MB. So we may have a max size issue in that reporting.  It has some crazy long link urls in it. That are hidden! You can see the page rendering stop as it loads each link.

I tried testing http://www.meishin-tire.com. It was 21MB and it looks like most of that is white space.

I'd say the data is off and we would need to look into how it is gathered to get more accurate data next time. What we could probably say is the we found over 10 sites where their initial html was larger than 16MB.

When scripting is unsupported or turned off in the browser, the `noscript` element helps to add an HTML section within a page. Considering the above script numbers, we were curious about the `noscript` element as well. 

Following the analysis, we found that about 49% of pages are using a `noscript` element. At the same time, about 16% of `noscript` elements were containing an `iframe` with a `src` value referring to "googletagmanager.com".

This seems to confirm the theory that the total number of `noscript` elements in the wild may be affected by common scripts like Google Tag Manager which enforce users to add a `noscript` snippet after the `body` start tag on a page.

What `type` attribute values are used with `script` elements?

{# TODO(authors, analysts): Should this be a figure? #}
- `text/javascript`: 60.03%
- `application/ld+json`: 1.68%
- `application/json`: 0.41%
- `text/template`: 0.41%

> - `text/html` 0.27% (!)

Author: Your comment showed how much nuance there is between placing the mark before or after the stat. I obviously misunderstood what you were trying to convey, so we should assume readers would too.

> When it comes to loading JavaScript module scripts using `type="module"`, we found that 0.13% of `script` elements currently specify this attribute-value combination. `nomodule` is used by 0.95% of all tested pages. Note that one metric relates to elements, the other to pages.

Reviewer 2: ```suggestion
When it comes to loading [JavaScript module scripts](https://jakearchibald.com/2017/es-modules-in-browsers/) using `type="module"`, we found that 0.13% of `script` elements currently specify this attribute-value combination. `nomodule` is used by 0.95% of all tested pages. Note that one metric relates to elements, the other to pages.
```

<figcaption>{{ figure_link(caption="Obsolete elements with more than 10,000 uses.", sheets_gid="1972617631", sql_file="pages_element_count_by_device_and_obsolete_elements.sql") }}</figcaption>
</figure>

Why are these still alive? Even `spacer` is still being used 1,584 times, and present on every 5,000th page. We know that Google has been using a `center` element on [their homepage](https://www.google.com/) [for 22 years](https://web.archive.org/web/19981202230410/https://www.google.com/) now, but why are there so many imitators?

#### `isindex`

> {# TODO(authors): What is the significance of this section? There are no new stats provided. #}

Reviewer 2: "was present once", i.e. the new stat is 1 page used `isindex`. I'll leave it to the authors to decide the noteworthyness of this; seems like we could clarify the wording if we are to keep this section.

Author: Gotcha, I read that as "there was a time that `isindex` existed" not that "`isindex` is used on exactly one page". Agree it's worth clarifying.

> If you were wondering: The [`isindex`](https://www.w3.org/TR/html401/interact/forms.html#edef-ISINDEX) element was present once. It was part of the HTML specs until [version 4.01](https://meiert.com/en/indices/html-elements/), and of XHTML in 1.0, yet only properly [specified](https://lists.w3.org/Archives/Public/public-whatwg-archive/2006Feb/0111.html) in 2006, and [removed](https://github.com/whatwg/html/pull/1095) in 2016.

Reviewer 2: ```suggestion
If you were wondering: The total number of [`isindex`](https://www.w3.org/TR/html401/interact/forms.html#edef-ISINDEX) elements in this dataset is: one. Exactly one page used an `isindex` element. It was part of the specs until [HTML 4.01 and XHTML 1.0](https://meiert.com/en/indices/html-elements/), yet only properly [specified](https://lists.w3.org/Archives/Public/public-whatwg-archive/2006Feb/0111.html) in 2006 (aligning with how it was implemented in browsers), and then [removed](https://github.com/whatwg/html/pull/1095) in 2016.
```

### Proprietary and made-up elements

> In our set of elements we found some that were neither standard HTML (nor SVG nor MathML) elements, nor custom ones, nor obsolete ones, but somewhat proprietary ones. The top 10 that we identified were the following:

Reviewer 4: 1. “were” → “are” (twice)
2. “top” → “Top“ (preference—I’m used to this being capitalized, and I find it clearer)

| `rel` | 171,802,460 | 1.96% |
| `value` | 140,666,779 | 1.61% |

<figcaption>{{ figure_link(caption="Top 10 attributes by frequency of use.", sheets_gid="1348855449", sql_file="pages_almanac_by_device_and_attribute_name_frequency.sql") }}</figcaption>
</figure>

The most popular attribute is `class`, with nearly 3 billion occurrences in our dataset and constituting 34% of all attributes in use. `class` is by far the most prevalent attribute.

> The `value` attribute, which specifies the value of an `input` element, surprisingly completes the top ten. It's surprising to us because, subjectively, we didn't get the impression values were used that frequently.

Reviewer 4: Please markup “values” as code, but like “`value`s” (I think we need to for clarity, as values without code markup would seem to apply to any type of value).

### Attributes on pages

Are there attributes that we find in every document? Not quite, but almost:

> <figure markdown>

Reviewer 4: The original draft was deliberate about text form as the information here didn’t seem that critical. If you object against the text form, please shorten this table to 10 attributes—it’s very little value for this much space otherwise.

Element | Pages (%)
-- | --
href | 99.21%
src | 99.18%
content | 98.88%
name | 98.61%
type | 98.55%
class | 98.24%
rel | 97.98%
id | 97.46%
style | 95.95%
alt | 90.75%
title | 85.50%
target | 82.98%
async | 81.96%
lang | 79.10%
width | 78.99%
height | 76.36%
charset | 75.44%
value | 67.57%
role | 62.97%
action | 62.34%
http-equiv | 62.30%
method | 60.27%
media | 60.14%
property | 55.68%
placeholder | 54.21%

<figcaption>{{ figure_link(
  caption="Top 25 attributes by page.",
  sheets_gid="1185369559",
  sql_file="pages_almanac_by_device_and_attribute_name_present.sql"
)}}</figcaption>
</figure>

> These results raise some questions that we cannot answer. For example, `type` is used on other elements too, but why this tremendous popularity? Especially given that it's usually not needed to specify for style sheets or scripts, with CSS and JavaScript being assumed default. Or, how do we really fare with `alt`? Do those 9.25% of pages not contain any images or are they just inaccessible? The rest of these 25 attributes make up all of those found on at least half of all pages.

Reviewer 4: This will need editing after the table got shortened. It may suffice to remove the last sentence, “The rest of these 25 attributes make up all of those found on at least half of all pages.”.

### `data-*` attributes

> Per the HTML spec,[`data-*` attributes](https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes) "are intended to store custom data, state, annotations, and similar, private to the page or application, for which there are no more appropriate attributes or elements." How are they used? What are the popular ones? Is there anything interesting here?

Reviewer 4: Please add a space after the first comma.

> The most popular ones stand out because they are almost twice as popular than each of the attributes that followed (with >1% use):

Reviewer 4: Please add “two”: “The two most popular ones”—clearer.

<figcaption>{{ figure_link(caption="The most popular <code>data-*</code> attributes.", sheets_gid="764700773", sql_file="pages_almanac_by_device_and_data_attribute_name_frequency.sql") }}</figcaption>
</figure>

Attributes like `data-type`, `data-id`, and `data-src` can have multiple generic uses although `data-src` is used a lot with lazy image loading via JavaScript (e.g., Bootstrap 4). [Bootstrap](https://getbootstrap.com/) again explains the presence of `data-toggle`, where it's used as a state styling hook on toggle buttons. The [Slick carousel plugin](https://kenwheeler.github.io/slick/) is the source of `data-slick-index`, whereas `data-element_type` is part of [Elementor's WordPress website builder](https://elementor.com/). Both `data-requiremodule` and `data-requirecontext`, then, are part of [RequireJS](https://requirejs.org/).

> Interestingly, the use of native lazy loading on images is similar to that of `data-src`. 3.86% of `img` elements use the `loading` attribute with a value of `lazy`. This appears to be growing very fast, as back in February, this number was about [0.8%](https://twitter.com/zcorpan/status/1237016679667970050). It's possible that these are being used together for a [cross-browser solution](https://addyosmani.com/blog/lazy-loading/).

Reviewer 2: Hmm, I thought the 3.86% number was of pages using `<img loading=lazy>`, not of `img` elements. If it's not % of pages, the numbers are not comparable.

Author: cc @Tiggerito

Reviewer 3: @zcorpan is right. This data is on a per page bases. So pages that have an img with loading set to lazy.

Reviewer 5: So the "this appears to be growing fast..." sentence should be removed?

Author: I've corrected the phrasing of the stat itself but left a TODO for the authors to update their interpretation.

Reviewer 2: @iandevlin no it can stay, since both stats are of pages.

Attributes like `data-type`, `data-id`, and `data-src` can have multiple generic uses although `data-src` is used a lot with lazy image loading via JavaScript (e.g., Bootstrap 4). [Bootstrap](https://getbootstrap.com/) again explains the presence of `data-toggle`, where it's used as a state styling hook on toggle buttons. The [Slick carousel plugin](https://kenwheeler.github.io/slick/) is the source of `data-slick-index`, whereas `data-element_type` is part of [Elementor's WordPress website builder](https://elementor.com/). Both `data-requiremodule` and `data-requirecontext`, then, are part of [RequireJS](https://requirejs.org/).

{# TODO(authors): Update this interpretation given that the lazy loading stat is in terms of pages, not img elements. #}
Interestingly, the use of native lazy loading on images is similar to that of `data-src`. 3.86% of pages use the `<img loading="lazy">` attribute. This appears to be growing very fast, as back in February, this number was about [0.8%](https://twitter.com/zcorpan/status/1237016679667970050). It's possible that these are being used together for a [cross-browser solution](https://addyosmani.com/blog/lazy-loading/).

## Miscellaneous

> We've covered the use of HTML in general as well as the adoption of top elements and attributes. In this section, we're reviewing some of the special cases of viewports, favicons, buttons, inputs, and links. Specifically, we will also be looking at links that may still point to "http" URLs.

Reviewer 4: Last sentence, no—this is not what we’re doing. These intros are following a scheme. The easiest path here would be to stay close to the draft: “Too many links may still point to “http” URLs is one thing we note here.”

| `initial-scale=1,maximum-scale=1,user-scalable=no,width=device-width` | 352,972 | 5.56% |
| `initial-scale=1,maximum-scale=1,user-scalable=0,width=device-width` | 249,662 | 3.93% |
| `width=device-width` | 231,668 | 3.65% |

<figcaption>{{ figure_link(caption="<code>viewport</code> specifications, and lack thereof.", sheets_gid="1414206386", sql_file="summary_pages_by_device_and_viewport.sql") }}</figcaption>
</figure>

The results show that almost half of the pages we analyzed are using the typical viewport `content` value. Still, around 10% of mobile pages are entirely missing a proper `content` value for the viewport meta element, with the rest of them using an improper combination of `maximum-scale`, `minimum-scale`, `user-scalable=no`, or `user-scalable=0`.

<p class="note">
  For a while now, the Edge mobile browser allows users to zoom into a web page to <a href="https://blogs.windows.com/windows-insider/2017/01/12/announcing-windows-10-insider-preview-build-15007-pc-mobile/">at least 500%</a>, regardless of the zoom settings defined by a web page employing the viewport meta element.
</p>

### Favicons

> The state of favicons is fascinating. Favicons work with or without markup, for example some browsers would fall back to [looking at the domain root](https://realfavicongenerator.net/faq#why_icons_in_root), browsers accept favicons in several image formats, and they are also supported in several dozen sizes. In fact, some tools are reported to generate 45 favicon sizes! For example, one favicon tool, [realfavicongenerator.net](https://realfavicongenerator.net/), would return _37_ favicon sizes if requested to handle every case. As of this time of writing, there is an [open issue](https://github.com/whatwg/html/issues/4758) for the HTML spec to help improve the situation.

Reviewer 4: I get some of the editing here (break large sentences down etc.), however I find it reads, “choppy.”

For the sake of simplicity I suggest to re-establish the old version, but keep the edits of the first and last sentence. Meaning and flow changed much in the middle section.

| … | | |
| SVG | 5,328 | 0.08% |

<figcaption>{{ figure_link(caption="Common favicon formats.", sheets_gid="1930085905", sql_file="pages_almanac_by_device_and_favicon_image_type.sql") }}</figcaption>
</figure>

There are a couple of surprises in here:

* Support for other formats is there but ICO is still the go-to format for favicons on the web.
* JPG is a relatively popular favicon format even though it may not yield the best results (or a comparatively large file size) for many favicon sizes.
* WebP is twice as popular as SVG! This might change, however, with [SVG favicon support](https://caniuse.com/link-icon-svg) improving.

### Button and input types

> {# TODO(authors): Can you cite any examples of this discussion? And what do the findings say about button misuse? #}

Reviewer 6: I don’t know about discussions but people write and talk a lot about it.

Vadim Makeev 2019: https://www.smashingmagazine.com/2019/02/buttons-interfaces/
HTMHell 2020: https://www.htmhell.dev/20-close-buttons/
Marcy Sutton 2017: https://www.youtube.com/watch?v=8XjwDq9zG4I

Reviewer 7: Thanks for adding these helpful resources @matuzo and I'd like to add these to the conversation as well:

- Links, Buttons, Submits, and Divs, Oh Hell
  https://adrianroselli.com/2016/01/links-buttons-submits-and-divs-oh-hell.html

- Links are not buttons. Neither are DIVs and SPANs
  https://karlgroves.com/2013/05/14/links-are-not-buttons-neither-are-divs-and-spans

Author: I'd suggest picking the most representative example and link to it from the word "discussion". I can apply that change if you let me know which URL you'd prefer.

Reviewer 7: I'd suggest linking Adrian Roselli's write-up as it contains lots of updates and links to other similar resources too.

<figure markdown>
| Percentile | Buttons per page |
|---|---|
| 10 | 0 |
| 25 | 0 |
| 50 | 1 |
| 75 | 5 |
| 90 | 13 |

<figcaption>{{ figure_link(caption="Distribution of the number of buttons per page.", sheets_gid="309769153", sql_file="pages_markup_by_device_and_percentile.sql") }}</figcaption>
</figure>

Pages in the 10th and 25th percentiles contain no buttons at all, while a page in the 90th percentile contains 13 native `button` elements. In other words, 10% of pages contain 13 or more buttons.

### Link targets

The [anchor element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), or `a` element, links web resources together. In this section, we analyze the adoption of the protocols included in these link targets.

<figure markdown>

> | Protocol | Occurrences | Pages (%) |

Reviewer 4: General note (@rviscomi)—all those header changes look like thin ice to me. If these are changed they need scrutiny to make sure we don’t inadvertently change meaning. (We had some discussions related to this and had taken a lot of care during editing of the doc that the headers are right. I don’t have the time right now to check all those new header labels to verify they are still accurate—all I can say is that they’re a minefield and should not be changed lightly.)

Author: There were a few headings I didn't change for that reason, but it's worth a second look at the ones that I did change to make sure that they're correct.

Reviewer 2: This certainly isn't occurrences, but pages.
```suggestion
| Protocol | Pages | Pages (%) |
```

| … | | |
| SVG | 5,328 | 0.08% |

<figcaption>{{ figure_link(caption="Common favicon formats.", sheets_gid="1930085905", sql_file="pages_almanac_by_device_and_favicon_image_type.sql") }}</figcaption>
</figure>

There are a couple of surprises in here:

* Support for other formats is there but ICO is still the go-to format for favicons on the web.

> * JPG is a relatively popular favicon format even though it may not yield the best results (or a comparatively large file size) for many favicon sizes.

Reviewer 4: Please replace “file size” by “weight”—that avoids confusion with the other sizes (dimensions) we’re talking about.

* WebP is twice as popular as SVG! This might change, however, with [SVG favicon support](https://caniuse.com/link-icon-svg) improving.

### Button and input types

There has been a lot of [discussion](https://adrianroselli.com/2016/01/links-buttons-submits-and-divs-oh-hell.html) on buttons lately and how often they are misused. We looked into this to present findings on some of the native HTML buttons.

{{ figure_markup(
  caption="Percent of pages with button elements.",
  content="60.56%",
  classes="big-number",
  sheets_gid="410549982",
  sql_file="pages_markup_by_device.sql"
) }}

<figure markdown>

> | Button types | Occurrences | Percentage |

Reviewer 4: “Percentage” is confusing and, I believe, wrong here. I’m not sure whether “Page percentage”, as per the draft, is perfectly clear, but just “Percentage” does not work. @zcorpan, you seem to have a clearer view here, can you help out and advise on alternative wording?

Reviewer 2: Is it percentage of pages? If so, maybe "Percentage (pages)"

Reviewer 2: Yeah

Author: Is the corresponding [`sheets_gid`](https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/edit#gid=410549982) correct? I'm not able to match up these results with the sheet. 15.9M can't be right because there are only 6,347,640 mobile pages in the dataset. @j9t or @Tiggerito could you clarify where these values came from?

Reviewer 4: @catalinred, can you advise? I think this is due to [this section](https://docs.google.com/document/d/1m1YvDqWgZfqmcmAEkCnXncIJwsM1QCu6Rv7aPWjG1Ms/edit#heading=h.nm1lcnapqkhr) pulling from four sheets, and I had a similar problem tracing this down.

|---|---|---|
| `<button type="button">` | 15,926,061 | 36.41% |
| `<button>` without type | 11,838,110 | 32.43% |
| `<button type="submit">` | 4,842,946 | 28.55% |
| `<input type="submit" value="…">` | 4,000,844 | 31.82% |
| `<input type="button" value="…">` | 1,087,182 | 4.07% |
| `<input type="image" src="…">` | 322,855 | 2.69% |
| `<button type="reset">` | 41,735 | 0.49% |

<figcaption>{{ figure_link(caption="Adoption of button types.", sheets_gid="410549982", sql_file="pages_markup_by_device.sql") }}</figcaption>
</figure>

> Our analysis shows that about 60% of pages contain a button element and more than half of the pages (32.43%) have at least one button that fails to specify a `type` attribute. Note that the button element has a [default type](https://dev.w3.org/html5/spec-LC/the-button-element.html) of `submit`, so the default behavior of buttons on these 32% of pages is to submit the current form data. To avoid possibly unexpected behavior like this, a best practice is to specify the type attribute.

Reviewer 4: 1. Please mark “type” (attribute) as code: “`type` attribute”

2. ⚠️ We need to be really specific here about what “button element” means. If we mean the `button` element, we should write that (and mark “button” as code), if we mean button elements in general, we should leave out code markup, but perhaps add “…in general”, or even call out `button` and `input` button elements.

@catalinred, can you comb through this section and check whether wording and markup is precise?

Reviewer 7: I double-checked [this sheet](https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/edit?ts=5f4e700c#gid=410549982) and the elements mentioned here are proper `button` elements only.

Here's how the phrase with backticks looks like, I hope this helps to clear the confusion:

Our analysis shows that about 60% of pages contain a `button` element and more than half of the pages (32.43%) have at least one `button` that fails to specify a `type` attribute. Note that the `button` element has a [default type](https://dev.w3.org/html5/spec-LC/the-button-element.html) of `submit`, so the default behavior of `button`s on these 32% of pages is to submit the current form data. To avoid possibly unexpected behavior like this, a best practice is to specify the `type` attribute.

Reviewer: Nit, but could we change to "more than half of **those** pages" as it looks weird saying half of pages with a % that isn't nearer 50%? Also should we use the same % value (32.43%) rather than shorten to 32% in the second instance (which makes it look like a different stat)?

Suggested text:

> Our analysis shows that about 60% of pages contain a `button` element and more than half of those pages (32.43%) have at least one `button` that fails to specify a `type` attribute. Note that the `button` element has a [default type](https://dev.w3.org/html5/spec-LC/the-button-element.html) of `submit`, so the default behavior of `button`s on these 32.43% of pages is to submit the current form data. To avoid possibly unexpected behavior like this, a best practice is to specify the `type` attribute.

Getting really niggly now! 😀

Reviewer 7: Sure, it sounds good to me!

<figure markdown>
| Percentile | Buttons per page |
|---|---|

> | 10 | 0 |

Reviewer 4: In my mind this is not adding precision—am I missing something or could it be that 25/0 is quite sufficient?

Reviewer 2: I prefer having it in for symmetry/completeness.

Reviewer 2: @rviscomi I think @j9t meant whether this row (10th percentile) is necessary

Author: Oh ok. +1 for completeness.

> | 10 | 0 |

Reviewer 4: In my mind this is not adding precision—am I missing something or could it be that 25/0 is quite sufficient?

Elements per page and element types per page stayed roughly the same, showing [no significant change](#element-diversity) in our HTML writing practice when compared to 2019. Such changes may require more time to manifest.

Proprietary product-specific elements like `g:plusone` (used on 17,607 pages in the mobile sample) and `fb:like` (11,335) have almost disappeared after still being [among the most popular ones](../2019/markup#products-and-libraries-and-their-custom-markup) last year. However, we observe more [custom elements](#custom-elements) for things like Slider Revolution, AMP, and Angular. Elements like `ym-measure`, `jdiv`, and `ymaps` are also still prevalent. What we imagine we're seeing here is that, under the sea of slowly changing practices, HTML is very much being developed and maintained, as authors toss deprecated markup and embrace new solutions.

We're leaving this open to you. What are your observations? What has caught your eye? What do you think has taken a turn for the worse, and what has improved? [Leave a comment](https://discuss.httparchive.org/t/2039) to share your thoughts!

## Conclusion

> When we don't need to [cover 14 years](../2019/markup) for analysis but only 1, one could almost get the impression that HTML is rather inert, that not much changes.

Reviewer: I think this sentence needs rewording. It reads really confusing to me.

Even ignoring the phrasing, I think it's a lot to ask readers to remember or understand what the 14 years stats is about to make sense of this sentence.You don't even mention it in this chapter. I know you link to it but you shouldn't have to click a link to understand what

Is this what you're trying to say? In which case I think we can jump straight into the next sentence without a paragraph break:

"The 2019 Web Almanac Markup chapter had 14 years of catch up to do since the last major study on the topic, so you'd think we wouldn't have much to cover in the year since. Yet we observe..."

Reviewer 4: SGTM.

<figcaption>{{ figure_link(caption="Blank relationships.", sheets_gid="1876528165", sql_file="pages_wpt_bodies_by_device.sql") }}</figcaption>
</figure>

As a rule of thumb and for [usability reasons](https://www.nngroup.com/articles/new-browser-windows-and-tabs/), prefer not to use `target="_blank"`. 

<p class="note">Within the latest Safari and Firefox versions, setting <code>target="_blank"</code> on <code>a</code> elements implicitly provides the same <code>rel</code> behavior as setting <code>rel="noopener"</code>. This is currently being <a href="https://bugs.chromium.org/p/chromium/issues/detail?id=898942">implemented in Chromium</a> as well.</p>

{# TODO(editors, authors): The following two sections each read like conclusions. Consider merging. #}
## Status and trends

We've sprinkled some observations throughout the chapter, and you'll have made your own observations. At the end of this 2020 analysis, here are some things that stood out for us.

> {# TODO(authors): Where does 3.97% come from? Can we link to it? #}

Reviewer 4: See doctypes section—however, on checking, I believe it’s 3.01% (100% – 96.99%). @zcorpan, am I missing something?

Reviewer 2: See analysis in https://docs.google.com/document/d/1m1YvDqWgZfqmcmAEkCnXncIJwsM1QCu6Rv7aPWjG1Ms/edit?usp=drivesdk&disco=AAAAIwTxu-k

Short version: some doctypes trigger quirks mode, so you need to count them, too.

Author: This is worth calling out in a standalone "big number" figure if significant. That will also provide a way to give more context and link it back to the data.

I see in the linked comment thread that the number was arrived at by copying the data to a new spreadsheet for analysis. I'd encourage you to keep all analysis in the public doc so it's easily discoverable. @zcorpan do you have edit access to do the same math in the official results sheet? (If not, request access through Sheets and I'll grant it shortly)

Reviewer 2: Done: https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/edit?ts=5f4e700c&pli=1#gid=1981441894&range=F102

Author: Thanks, added a big number figure and referenced the sheet.

<figcaption>{{ figure_link(caption="Blank relationships.", sheets_gid="1876528165", sql_file="pages_wpt_bodies_by_device.sql") }}</figcaption>
</figure>

> As a rule of thumb and for [usability reasons](https://www.nngroup.com/articles/new-browser-windows-and-tabs/), prefer not to use `target="_blank"`. 

Reviewer 4: Please add “in the first place” at the end.

<p class="note">Within the latest Safari and Firefox versions, setting <code>target="_blank"</code> on <code>a</code> elements implicitly provides the same <code>rel</code> behavior as setting <code>rel="noopener"</code>. This is already <a href="https://chromium-review.googlesource.com/c/chromium/src/+/1630010">implemented in Chromium</a> as well and will land in Chrome 88.</p>

> {# TODO(editors, authors): The following two sections each read like conclusions. Consider merging. #}

Reviewer 4: I also feel something still needs to be done here. Merging the section may require a bigger rewrite… Is there something elegant we could do with the heading here, and rename it from “Status and trends” to just “Trends”? “Observations”?

Reviewer: I tend to agree with Rick that I'm confused with this section. It seems to be repeating some stats highlighted previously, yet we have a separate Conclusion where you'd normally expect a summary to be.

I'd merge it, but also move the last section "We're leaving this open to you. What are your observations? ... Leave a comment to share your thoughts!" to the very end. It's a nice way to hand over the chapter to the user and also leads on to the new "Explore the results" heading.

If you merge the first two paragraphs of the current Conclusion ("When we don't need to..." and "Yet what we observe with this year's data"), I think they lead on quite nicely from the "Proprietary product-specific elements..." paragraph in this section, so it reads as one summary conclusion. Slightly long, but that's OK.

Reviewer 4: Just playing with options, what speaks against “Observations”? It's not perfect yet, but it seems to be an improvement that allows to cover very specific notes in this former “Status” section, and shift to general commentary in the conclusion.

Author: You still end up with two conclusions by different names :/

Reviewer 4: …I think @bazzadp’s approach could work. I had to look at this [in document form](https://docs.google.com/document/d/1YR9HpqOPOCy7eYQC2IngzrqjPmYLSFoPCL_W8qyXJw4/edit#) to get a better feel for and play with it (paragraphs I edited for this are marked)… while I want to read through this again, what do you think?

## Status and trends

> We've sprinkled some observations throughout the chapter, and you'll have made your own observations. At the end of this 2020 analysis, here are some things that stood out for us.

Reviewer 4: The duplicate “observations“ strikes me now. Replace the latter by “taken your own notes” or something like that?

Reviewer: I'm struggling with this paragraph. It seems to say "We saw some things, you saw some things. Here's some things we saw."

I think we can simplify it for a start by cutting the middle clause since you cover it later with the final paragraph in this section:

> We're leaving this open to you. What are your observations? What has caught your eye? What do you think has taken a turn for the worse, and what has improved? Leave a comment to share your thoughts!

What about the following for the remaining pieces:

```suggestion
We've touched on some observations throughout the chapter, but as a reflection on the state of markup in 2020, here are some things that really stood out for us:
```

Reviewer 4: I like it. I’d prefer “some more things that stood out” over “some things that really stood out”, as we haven’t made comments on these things any earlier.

{{ figure_markup(
  caption="Percent of pages with a quirky doctype.",
  content="3.97%",
  classes="big-number",
  sheets_gid="1981441894",
  sql_file="summary_pages_by_device_and_doctype.sql"
) }}

{# TODO(authors): Changed Simon's quote to a paraphrase, since it's not clear which part is verbatim. If there's a quote, let's wrap it in quotes.  #}

> Fewer pages land in quirks mode. In 2016, that number was at [around 7.4%](https://discuss.httparchive.org/t/how-many-and-which-pages-are-in-quirks-mode/777). At the end of 2019, we observed [4.85%](https://twitter.com/zcorpan/status/1205242913908838400). And now, we're at about 3.97%. This trend, to paraphrase [Simon Pieters](./contributors#zcorpan) in his review of this chapter, seems indeed clear and encouraging.

Reviewer 4: Please remove “indeed”—I don’t think the reference is clear.

> Fewer pages land in quirks mode. In 2016, that number was at [around 7.4%](https://discuss.httparchive.org/t/how-many-and-which-pages-are-in-quirks-mode/777). At the end of 2019, we observed [4.85%](https://twitter.com/zcorpan/status/1205242913908838400). And now, we're at about 3.97%. This trend, to paraphrase [Simon Pieters](./contributors#zcorpan) in his review of this chapter, seems indeed clear and encouraging.

Reviewer 2: Did we want to mention the 3.97% quirks mode stat in the "Doctypes" section? Maybe put it in the table as "No doctype or any quirky doctype"

Although we lack historic data to draw the full development picture, "meaningless" `div`, `span`, and `i` markup has pretty much [replaced](#top-elements) the `table` markup we've observed in the 1990s and early 2000s. While one may question whether `div` and `span` elements are always used without there being a semantically more appropriate alternative, these elements are still preferable to `table` markup, though, as during the heyday of the old web, these were seemingly used for everything but tabular data.

Elements per page and element types per page stayed roughly the same, showing [no significant change](#element-diversity) in our HTML writing practice when compared to 2019. Such changes may require more time to manifest.

> Proprietary product-specific elements like `g:plusone` (used on 17,607 pages in the mobile sample) and `fb:like` (11,335) have almost disappeared after still being [among the most popular ones](../2019/markup#products-and-libraries-and-their-custom-markup) last year. However, we observe more [custom elements](#custom-elements) for things like Slider Revolution, AMP, and Angular. Elements like `ym-measure`, `jdiv`, and `ymaps` are still prevalent. What we imagine we're seeing here is that, under the sea of slowly changing practices, HTML is very much being developed and maintained, as authors toss deprecated markup and embrace new solutions.

Reviewer 4: Please add “also”—“are also still prevalent”. (Something to aid reading flow.)

We're leaving this open to you. What are your observations? What has caught your eye? What do you think has taken a turn for the worse, and what has improved? [Leave a comment](https://discuss.httparchive.org/t/2039) to share your thoughts!

## Conclusion

When we don't need to [cover 14 years](../2019/markup) for analysis but only 1, one could almost get the impression that HTML is rather inert, that not much changes.

> Yet what we observe with this year's data is that there's a lot of movement at the bottom and near the shore of said sea of HTML. We approach near-complete adoption of living HTML, are quick to prune our pages of fads like Google and Facebook widgets, and we're fast in adopting and shunning frameworks. Both Angular and AMP, though "component frameworks", seem to have significantly lost in popularity, likely for solutions like React and Vue.

Reviewer 4: I prefer the old structure, as it isolates the examples in parentheses, which seems to be clearer.

If you do want to fold the examples into the sentences, then keep with the logical units—adoption; fads; frameworks. That is, more like this:

“We approach near-complete adoption of living HTML. We are quick to prune our pages of fads like Google and Facebook widgets. We're also fast in adopting and shunning frameworks, as both Angular and AMP (though a "component framework") seem to have significantly lost in popularity, likely for solutions like React and Vue.”

Note: “component framework” relates only to AMP, and must therefore stay singular.

> And still, there are no signs we exhausted the options HTML gives us. The median of 30 different elements used on a given page, which is roughly a quarter of the elements HTML provides us with, suggests a rather one-sided use of HTML. That is supported by the immense popularity of elements like `div` and `span`, and no custom elements to potentially meet the demands that these two elements may represent. Unfortunately, we couldn't validate each document in the sample. However, anecdotally and to be taken with caution, we learned that [79%](https://github.com/HTTPArchive/almanac.httparchive.org/issues/899#issuecomment-717856201) of W3C-tested documents have validation errors. After everything we've seen, it looks like we're still far from mastering the craft of HTML.

Reviewer 4: 1. I don’t like the break and pause for validation (“…sample. However”)—it drags the whole point out for little value. Could you join these to form one unit?

2. Please replace “we’ve seen” by “we’re seeing” (consistency in tense).

Author: 1. Could you suggest the change you're looking to see?

2. Not consistent with "we learned" in the previous sentence. Should we also replace that or keep them both?

Reviewer 4: 1. I think a semicolon would help: “…sample; however…”.

2. Probably my suggested change isn’t that important; now I rather wonder we should instead go for “we’ve learned” 😄 You decide.

> That compels us to close with an appeal. Pay attention to HTML. Focus on HTML. It's important and worthwhile to invest in HTML. HTML is a document language that may not have the charm of a programming language, and yet the web is built on it. Use less HTML and learn what's really needed. Use more appropriate HTML—learn what's available and what it's there for. And [validate](https://validator.w3.org/docs/why.html) your HTML. Anyone can write invalid HTML (just invite the next person you meet to write an HTML document and validate the output) but a professional developer can be expected to produce valid HTML. Writing correct and valid HTML is a craft to take pride in.

Reviewer 4: This happens throughout the doc, but here I find the change from uppercase “Web” to lowercase “web” most dramatic. Which web is meant here? Yes, probably people guess which one. But _the Web_ is _the Web_, is a statement, is an exclamation mark. Here, at last, the effort to be consistently lowercase reduces the meaning and weight, perhaps of the whole chapter. (I added a bit of drama, on purpose, but really—HTML makes for the _Web_.)

Author: This is a conscious choice in our style guide: https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Editors'-Guide#user-content-technical:~:text=Use%20lowercase%20%22web%22

Do you feel strongly that this usage is significant enough compared to lowercase phrases like "web page" or "web developer" to break that colloquialism?

Reviewer 4: Personally, yes, very much! I think there’s the Web, one Web, and then many websites, web apps, web developers, &c. (I never got why one would write “web” lowercase—I think “which web?” then.) But that’s just me; to make the point I think at least here it should be “Web”; and if that is a problem and we need to game the style guide, then maybe “World Wide Web”? But there I’m not sure we need to go there.

Reviewer: I had the same issue when writing my book I used Web and Internet but publishers insisted on web and internet. I’ve come round to lowercase and prefer it now but it took a while to persuade me. Apparently lowercase is more and common and the capitalisation is seen as somewhat old fashioned nowadays in style guides: https://www.wired.com/2016/04/ap-finally-realizes-2016-will-let-us-stop-capitalizing-internet/

Author: Given that we use "web" lowercase even outside of pairings like "web page" and "web developer", I'm inclined to keep this consistent. For example, the home page prominently refers to this as "The 2020 state of the web report". It'd be strangely inconsistent to make it a proper noun in this one instance. Regarding "gaming" the style guide, we _explicitly discourage_ "World Wide Web"! 😄 

This is more of a question about overall tone and formality, which goes beyond just this chapter, similar to the quotes thing. Since both Web and web are accepted and we've standardized on the latter, I think we should stick to it for now and reevaluate later if necessary.

PS I got a laugh out of this quote from the Wired article:

> There is no earthly reason to capitalize any of these words. Actually there never was.

Reviewer: I think we should resolve this. We need consistency between all the chapters and we've decided to use lowercase form for the Web Almanac. I get that others (including the authors here) may prefer the other way, but we need the consistency so that isn't something we can resolve here.

@j9t please raise a separate issue for discussion and visibility to the rest of the contributors if you still feel that strongly about this (like I did for smart quotes). Also means we can refer to it, if this item comes up again in another chapter - especially if we as a team decide to change to capitalised form.
